Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm

```diff
@@ -0,0 +1,271 @@
+
+    if (syncStorageProperties.get().contains(propertyName))
+        return m_type == StorageType::Sync;
+
+    if (propertyName == "setAccessLevel"_s)
+        return m_type == StorageType::Session;
+
+    ASSERT_NOT_REACHED();
+    return false;
+}
+
+void WebExtensionAPIStorageArea::get(NSObject *items, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea/get
+
+    auto returnEmptyResult = ^() {
+        callback->call(@{ });
+    };
+
+    auto validateJSONAndReturnData = ^(id items, JSONOptionSet options, NSString **outExceptionString) {
+        if (!isValidJSONObject(items, options)) {
+            *outExceptionString = toErrorString(nil, @"items", @"it is not JSON-serializable");
+            return;
+        }
+
+        NSData *data = encodeJSONData(items, options);
+        if (!data) {
+            *outExceptionString = toErrorString(nil, @"items", @"it is not JSON-serializable");
+            return;
+        }
+
+        // FIXME: <https://webkit.org/b/264892> Get data from storage.
+        callback->call();
+    };
+
+    if ([items isKindOfClass:NSDictionary.class]) {
+        auto *keysWithDefaultValues = (NSDictionary *)items;
+        if (!keysWithDefaultValues.count) {
+            returnEmptyResult();
+            return;
+        }
+
+        validateJSONAndReturnData(keysWithDefaultValues, { }, outExceptionString);
+        return;
+    }
+
+    if ([items isKindOfClass:NSArray.class]) {
+        auto *keys = (NSArray *)items;
+        if (!keys.count) {
+            returnEmptyResult();
+            return;
+        }
+
+        validateJSONAndReturnData(keys, { JSONOptions::FragmentsAllowed }, outExceptionString);
+        return;
+    }
+
+    if ([items isKindOfClass:NSString.class]) {
+        NSString *key = (NSString *)items;
+        if (!key.length) {
+            returnEmptyResult();
+            return;
+        }
+
+        validateJSONAndReturnData(key, { JSONOptions::FragmentsAllowed }, outExceptionString);
+        return;
+    }
+
+    if (items && ![items isKindOfClass:NSNull.class]) {
+        *outExceptionString = *outExceptionString = toErrorString(nil, @"items", @"an invalid parameter was passed");
+        return;
+    }
+}
+
+void WebExtensionAPIStorageArea::getBytesInUse(NSObject *items, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea/getBytesInUse
+
+    bool isString = [items isKindOfClass:NSString.class];
+    bool isArray = [items isKindOfClass:NSArray.class];
+    bool isNull = [items isKindOfClass:NSNull.class];
+
+    if (items && !isString && !isArray && isNull) {
+        *outExceptionString = toErrorString(nil, @"items", @"it is expected to be either a string, array, or null");
+        return;
+    }
```

Comment:
To keep the error consistent, use `validateObject`.

```
if (items && !validateObject(items, @"keys", [NSOrderedSet orderedSetWithObjects:NSString.class, NSArray.class, NSNull.class, nil], outExceptionString))
    return;
```

Reply:
forgot i could use NSOrderedSet for multiple input types! makes this much simpler. thanks!